distributed/deploy/cluster.py

```diff
@@ -24,45 +26,84 @@
-    common methods and an IPython widget display.
+    This class contains common functionality for Dask Cluster manager classes.
 
-    Clusters inheriting from this class should provide the following:
+    To implement this class, you must provide
 
-    1.  A local ``Scheduler`` object at ``.scheduler``
-    2.  scale_up and scale_down methods as defined below::
+    1.  A ``scheduler_comm`` attribute, which is a connection to the scheduler
+        following the ``distributed.core.rpc`` API.
+    2.  Implement ``scale``, which takes an integer and scales the cluster to
+        that many workers, or else set ``_supports_scaling`` to False
 
-        def scale_up(self, n: int):
-            ''' Brings total worker count up to ``n`` '''
+    For that, should should get the following:
 
-        def scale_down(self, workers: List[str]):
-            ''' Close the workers with the given addresses '''
-
-    This will provide a general ``scale`` method as well as an IPython widget
-    for display.
+    1.  A standard ``__repr__``
+    2.  A live IPython widget
+    3.  Adaptive scaling
+    4.  Integration with dask-labextension
+    5.  A ``scheduler_info`` attribute which contains an up-to-date copy of
+        ``Scheduler.identity()``, which is used for much of the above
+    6.  Methods to gather logs
+    """
 
-    Examples
-    --------
+    _supports_scaling = True
 
-    >>> from distributed.deploy import Cluster
-    >>> class MyCluster(cluster):
-    ...     def scale_up(self, n):
-    ...         ''' Bring the total worker count up to n '''
-    ...         pass
-    ...     def scale_down(self, workers):
-    ...         ''' Close the workers with the given addresses '''
-    ...         pass
+    def __init__(self, asynchronous):
+        self.scheduler_info = {}
+        self.periodic_callbacks = {}
+        self._asynchronous = asynchronous
 
-    >>> cluster = MyCluster()
-    >>> cluster.scale(5)                       # scale manually
-    >>> cluster.adapt(minimum=1, maximum=100)  # scale automatically
+        self.status = "created"
 
-    See Also
-    --------
-    LocalCluster: a simple implementation with local workers
-    """
-
-    def adapt(self, Adaptive=Adaptive, **kwargs):
+    async def _start(self):
+        comm = await self.scheduler_comm.live_comm()
+        await comm.write({"op": "subscribe_worker_status"})
+        self.scheduler_info = await comm.read()
+        self._watch_worker_status_comm = comm
+        self._watch_worker_status_task = asyncio.ensure_future(
+            self._watch_worker_status(comm)
+        )
+        self.status = "running"
+
+    async def _close(self):
+        await self._watch_worker_status_comm.close()
+        await self._watch_worker_status_task
+
+        for pc in self.periodic_callbacks.values():
+            pc.stop()
+        self.scheduler_comm.close_rpc()
+
+        self.status = "closed"
+
+    def close(self, timeout=None):
```

Comment:
Should `_close`/`close` be idempotent? I think right now it is just because all the operations in `_close` happen to be idempotent themselves, but it would maybe be good to be explicit with this.

Reply:
I've added a check to pass if the status is closed.  However there are still cases where this gets called twice that could be problematic.